File: python/extended_camera_model/utils/camera_model.py
# Copyright (C) 2024 twyleg, Marvin-VW
import numpy as np
import cv2 as cv
from typing import List, Tuple

class CameraModel:
    def __init__(self, sensor_width: float, sensor_height: float, focal_length: float, resolution_x: int, resolution_y: int, u0: int, v0: int):
        self.sensor_width: float = sensor_width
        self.sensor_height: float = sensor_height
        self.focal_length: float = focal_length
        self.resolution_x: int = resolution_x
        self.resolution_y: int = resolution_y
        self.u0: int = u0
        self.v0: int = v0

        self.camera_image = np.zeros((resolution_y, resolution_x, 3), dtype=np.uint8)
        self.reset_camera_image()

        # Create camera transformation matrix I_T_C
        rho_width = sensor_width / resolution_x
        rho_height = sensor_height / resolution_y

        matrix_k = np.array([
            [1/rho_width,   0,              u0],
            [0,             1/rho_height,   v0],
            [0,             0,               1],
        ])

        matrix_c = np.array([
            [focal_length,  0,              0,  0],
            [0,             focal_length,   0,  0],
            [0,             0,              1,  0],
        ])

        self.I_T_C = np.matmul(matrix_k, matrix_c)

    
    @staticmethod
    def transform_normals_to_world_space(normals, V_T_Cube):
        normals_in_world_space = V_T_Cube[:3, :3] @ normals
        return normals_in_world_space
    

    @staticmethod
    def world_transform(triangle, V_T_Cube):
        transformed_triangles = []

        for point in triangle:
                transformed_triangle = V_T_Cube @ point
                transformed_triangles.append(transformed_triangle)

        return transformed_triangles
    
    @staticmethod
    def camera_transform(object, C_T_V):
        transformed_triangles = []

        for point in object:
            transformed_triangle = tuple(C_T_V @ point)
            transformed_triangles.append(transformed_triangle)

        return transformed_triangles
    
    
    def draw_all_cube_points(self, triangles: List) -> None:

        for triangle in triangles:
            for point in triangle.camera_points:
                self.draw_camera_image_point(point)


    def draw_camera_image_point(self, C_point: np.array) -> None:
        I_point = np.matmul(self.I_T_C, C_point)
        u = int(I_point[0] / I_point[2])
        v = int(I_point[1] / I_point[2])
        cv.circle(self.camera_image, (u, v), 5, (255, 0, 0), 2)
    
    def draw_all_cube_lines(self, triangles : List) -> None:

        for triangle in triangles:
            for i in range(3):
                C_point0 = triangle.camera_points[i]
                C_point1 = triangle.camera_points[(i + 1) % 3]
                self.draw_camera_image_line(C_point0, C_point1)


    def draw_camera_image_line(self, C_point0: np.array, C_point1: np.array) -> None:
        I_point0 = np.matmul(self.I_T_C, C_point0)
        I_point1 = np.matmul(self.I_T_C, C_point1)

        u0 = int(I_point0[0] / I_point0[2])
        v0 = int(I_point0[1] / I_point0[2])

        u1 = int(I_point1[0] / I_point1[2])
        v1 = int(I_point1[1] / I_point1[2])

        cv.line(self.camera_image, (u0, v0), (u1, v1), (0, 0, 0), 1)

    def draw_camera_image_arrow(self, C_point0: np.array, C_point1: np.array) -> None:
        try:
            I_point0 = np.matmul(self.I_T_C, C_point0)
            I_point1 = np.matmul(self.I_T_C, C_point1)

            u0 = int(I_point0[0] / I_point0[2])
            v0 = int(I_point0[1] / I_point0[2])

            u1 = int(I_point1[0] / I_point1[2])
            v1 = int(I_point1[1] / I_point1[2])

            cv.arrowedLine(self.camera_image, (u0, v0), (u1, v1), (0, 255, 0), 2)
        except:
            raise ValueError(f"Could draw normal {C_point0}, {C_point1}")


    def fill_cube_faces(self, triangles) -> None:
        for triangle in triangles:
            I_points = []

            for C_point in triangle.camera_points:
                I_point = np.matmul(self.I_T_C, C_point)
                
                u = int(I_point[0] / I_point[2])
                v = int(I_point[1] / I_point[2])

                I_points.append((u, v))
            
            Poly_Points = np.array(I_points, np.int32)
            cv.fillPoly(self.camera_image, [Poly_Points], triangle.color)

    def draw_poly(self, points):

        I_points = []

        for point in points:

            I_point = np.matmul(self.I_T_C, point)
                
            u = int(I_point[0] / I_point[2])
            v = int(I_point[1] / I_point[2])

            I_points.append((u, v))
            
        Poly_Points = np.array(I_points, np.int32)
        hull = cv.convexHull(Poly_Points)
        cv.fillPoly(self.camera_image, [hull], (50,50,50))


    def reset_camera_image(self) -> None:
        self.camera_image.fill(255)


    @staticmethod
    def get_camera_vectors(V_T_C: np.array) -> np.array:
        rotation_matrix = V_T_C[:3, :3]
        forward_vector = -rotation_matrix[:, 2]
        camera_position = V_T_C[:3, 3]
        final_vector_x = forward_vector[0] + camera_position[0]
        final_vector_y = forward_vector[1] + camera_position[1]
        final_vector_z = forward_vector[2] + camera_position[2]
        return (final_vector_x, final_vector_y, final_vector_z)

File: cpp/extended_camera_model/libs/playground_camera_model/core/Camera/CameraModel.cc
// Copyright (C) 2024 twyleg
#include "CameraModel.h"
#include "Shape.h"
#include <opencv2/opencv.hpp>

CameraModel::CameraModel(double sensorWidth, double sensorHeight, double focalLength, uint32_t resolutionX, uint32_t resolutionY, uint32_t u0, uint32_t v0) {
    this->sensorWidth = sensorWidth;
    this->sensorHeight = sensorHeight;
    this->focalLength = focalLength;
    this->resolutionX = resolutionX;
    this->resolutionY = resolutionY;
    this->u0 = u0;
    this->v0 = v0;

    cameraImage.create(resolutionY, resolutionX, CV_8UC3);
    cameraImage = cv::Scalar(255, 255, 255);

    const double rhoWidth = sensorWidth / resolutionX;
    const double rhoHeight = sensorHeight / resolutionY;

    double matrixK_data[3 * 3] = {
        (1 / rhoWidth), 0, static_cast<double>(u0),
        0, (1 / rhoHeight), static_cast<double>(v0),
        0, 0, 1
    };

    double matrixC_data[3 * 4] = {
        focalLength, 0, 0, 0,
        0, focalLength, 0, 0,
        0, 0, 1, 0
    };

    const cv::Mat matrixK(3, 3, CV_64F, matrixK_data);
    const cv::Mat matrixC(3, 4, CV_64F, matrixC_data);

    I_T_C.create(3, 4, CV_64F);
    I_T_C = matrixK * matrixC;
}

void CameraModel::drawAllPoints(const std::vector<triangle>& mesh) {
    for (const auto& tri : mesh) {
        for (const auto& vertex : tri.point) {
            drawCameraImagePoint(vertex);
        }
    }
}

void CameraModel::drawCameraImagePoint(const cv::Mat& C_point){

	const cv::Mat I_point = I_T_C * C_point;

	const int32_t u = I_point.at<double>(0) / I_point.at<double>(2);
	const int32_t v = I_point.at<double>(1) / I_point.at<double>(2);

	const cv::Point point(u,v);

	cv::circle(cameraImage, point, 5, cv::Scalar(255,0,0), 2);

}

void CameraModel::drawAllLines(const std::vector<triangle>& mesh) {
    for (const auto& tri : mesh) {
        drawCameraImageLine(tri.point[0], tri.point[1]);
        drawCameraImageLine(tri.point[1], tri.point[2]);
        drawCameraImageLine(tri.point[2], tri.point[0]);
    }
}


void CameraModel::drawCameraImageLine(const cv::Mat& C_point0, const cv::Mat& C_point1){

	const cv::Mat I_point0 = I_T_C * C_point0;
	const cv::Mat I_point1 = I_T_C * C_point1;

	const int32_t u0 = I_point0.at<double>(0) / I_point0.at<double>(2);
	const int32_t v0 = I_point0.at<double>(1) / I_point0.at<double>(2);

	const int32_t u1 = I_point1.at<double>(0) / I_point1.at<double>(2);
	const int32_t v1 = I_point1.at<double>(1) / I_point1.at<double>(2);

	const cv::Point point0(u0,v0);
	const cv::Point point1(u1,v1);

	cv::line(cameraImage, point0, point1, cv::Scalar(255,0,0), 1);
}

void CameraModel::transform(const cv::Mat* matrix, const std::vector<triangle>& original_mesh, std::vector<triangle>& new_mesh) {
    new_mesh.clear();
    
    // Iterate over each triangle in the original mesh
    for (const auto& tri : original_mesh) {
        triangle new_tri;
        
        // Iterate over each vertex in the triangle
        for (int i = 0; i < 3; ++i) {
            // Apply the transformation matrix to the vertex
            cv::Mat point = tri.point[i];
            cv::Mat transformed_point = (*matrix) * point;

            // Update the new triangle with the transformed coordinates
            new_tri.point[i] = transformed_point;
        }
        
        // Add the transformed triangle to the new mesh
        new_mesh.push_back(new_tri);
    }
}



void CameraModel::resetCameraImage() {
    cameraImage = cv::Scalar(255, 255, 255);
}

cv::Mat& CameraModel::getCameraImage() {
    return cameraImage;
}